Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -0,0 +1,307 @@
+// `GrpcStream`. In turn, `GrpcStream` will decide whether to notify its
+// observer.
+//
+// `GrpcStream` owns the gRPC objects (such as `grpc::ClientContext`) that must
+// be valid until all `StreamOperation`s issued by this stream come back from
+// the gRPC completion queue. `StreamOperation`s contain an `std::promise` that
+// is fulfilled once the operation is taken off the gRPC completion queue, and
+// `StreamOperation::WaitUntilOffQueue` allows blocking on this. `GrpcStream`
+// holds non-owning pointers to all operations that it issued (and removes
+// pointers to completed operations). `GrpcStream::Finish` and
+// `GrpcStream::WriteAndFinish` block on `StreamOperation::WaitUntilOffQueue`
+// for all the currently-pending operations, thus ensuring that the stream can
+// be safely released (along with the gRPC objects the stream owns) after
+// `Finish` or `WriteAndFinish` have completed.
+
+namespace internal {
+
+StreamWrite* BufferedWriter::EnqueueWrite(grpc::ByteBuffer&& write) {
+  queue_.push(write);
+  return TryStartWrite();
+}
+
+StreamWrite* BufferedWriter::TryStartWrite() {
+  if (queue_.empty() || has_active_write_) {
+    return nullptr;
+  }
+
+  has_active_write_ = true;
+  grpc::ByteBuffer message = std::move(queue_.front());
+  queue_.pop();
+  return StreamOperation::ExecuteOperation<StreamWrite>(
+      stream_, call_, firestore_queue_, std::move(message));
+}
+
+StreamWrite* BufferedWriter::DequeueNextWrite() {
+  has_active_write_ = false;
+  return TryStartWrite();
+}
+
+}  // namespace internal
+
+using internal::BufferedWriter;
+
+GrpcStream::GrpcStream(
+    std::unique_ptr<grpc::ClientContext> context,
+    std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+    GrpcStreamObserver* observer,
+    AsyncQueue* firestore_queue)
+    : context_{std::move(context)},
+      call_{std::move(call)},
+      observer_{observer},
+      firestore_queue_{firestore_queue},
+      // Store the current generation of the observer.
+      generation_{observer->generation()},
+      buffered_writer_{this, call_.get(), firestore_queue_} {
+}
+
+GrpcStream::~GrpcStream() {
+  HARD_ASSERT(operations_.empty(),
+              "GrpcStream is being destroyed without proper shutdown");
+}
+
+void GrpcStream::Start() {
+  HARD_ASSERT(state_ == State::NotStarted, "Stream is already started");
+  state_ = State::Starting;
+  Execute<StreamStart>();
+}
+
+void GrpcStream::Read() {
+  HARD_ASSERT(!has_pending_read_,
+              "Cannot schedule another read operation before the previous read "
+              "finishes");
+  HARD_ASSERT(state_ == State::Open, "Read called when the stream is not open");
+
+  has_pending_read_ = true;
+  Execute<StreamRead>();
+}
+
+void GrpcStream::Write(grpc::ByteBuffer&& message) {
+  HARD_ASSERT(state_ == State::Open,
+              "Write called when the stream is not open");
+  BufferedWrite(std::move(message));
+}
+
+void GrpcStream::Finish() {
+  if (state_ == State::NotStarted) {
```

Comment:
Done.

The only thing that appears to be necessary is an `is_finishing_` flag, because it affects behavior. It's possible for several operations to fail, and it seems easier to only initiate the "finish due to error" sequence once.